firstvoices/backend/search/indexing.py

```diff
@@ -0,0 +1,185 @@
+        raise NotImplementedError()
+
+    @classmethod
+    def add_to_index(cls, instance):
+        try:
+            new_index_document = cls.create_index_document(instance)
+            new_index_document.save()
+            cls.refresh()
+        except ConnectionError as e:
+            log_connection_error(e, instance)
+        except Exception as e:
+            log_fallback_exception(e, instance)
+
+    @classmethod
+    def update_in_index(cls, instance):
+        try:
+            search_result = search_by_id(instance.id, cls.index)
+            existing_index_document = cls.document.get(id=search_result["_id"])
+
+            new_index_document = cls.create_index_document(instance)
+            new_values = new_index_document.to_dict(False, False)
+
+            existing_index_document.update(**new_values)
+
+            cls.refresh()
+
+        except ConnectionError as e:
+            log_connection_error(e, instance)
+        except NotFoundError as e:
+            log_not_found_error(e, instance)
+        except Exception as e:
+            log_fallback_exception(e, instance)
+
+    @classmethod
+    def remove_from_index(cls, instance):
+        try:
+            search_result = search_by_id(instance.id, cls.index)
+            existing_index_document = cls.document.get(id=search_result["_id"])
+            existing_index_document.delete()
+            cls.refresh()
+        except ConnectionError as e:
+            log_connection_error(e, instance)
+        except NotFoundError as e:
+            log_not_found_error(e, instance)
+        except Exception as e:
+            log_fallback_exception(e, instance)
+
+    @classmethod
+    def refresh(cls):
+        index = Index(ELASTICSEARCH_LANGUAGE_INDEX)
+        index.refresh()
+
+    @classmethod
+    def rebuild(cls):
+        es = connections.get_connection()
+        current_index = cls._get_current_index(es)
+        new_index = cls._create_new_write_index()
+
+        try:
+            cls._add_all(es)
+
+        except Exception as e:
+            # If we are not able to complete the new index, delete it and leave the current one as read + write alias
+            logger = logging.getLogger(ELASTICSEARCH_LOGGER)
+            logger.error(
+                "The following error occurred while adding documents to index [%s]. \
+                Deleting new index, making current index default. "
+                % cls.index
+            )
+            logger.error(e)
+            new_index.delete(ignore=404)
+            raise e
+
+        # Removing old index
+        if current_index:
+            current_index.delete(ignore=404)
+
+        cls._remove_write_alias(es, new_index)
+
+    @classmethod
+    def _get_current_index(cls, es_connection):
+        all_indices = es_connection.indices.get_alias(index="*")
+        related_indices = []
+        for index, value in all_indices.items():
+            if str(cls.index) in value["aliases"].keys():
```

Comment:
This condition is the one functional change from the other indexing-- matching on the alias dictionary rather than a substring in the index name